Installing-and-Configuring/Upgrading.md

```diff
@@ -18,11 +18,12 @@ performed via software, although you'll probably want a hardware programmer
 since we don't have a fail-safe recovery mechanism in the event of
 a bad flash or buggy firmware.
 
-Additionally, *reflashing the firmware will change the TPM PCRs*.
-This will require generating a new TPM TOTP token and a new drive
-encryption key.  Be sure you have your TPM owner's password and your
-disk encryption recovery key or passphrase available since, by design,
-the disk key is not accessible to the recovery shell.
+Additionally, *flashing a new firmware will change the TPM PCRs*. This will require generating a new TPM TOTP token and a new drive encryption key. Be sure you have your TPM owner’s password and your disk recovery key or passphrase available since, by design, the disk keys are not accessible to the recovery shell.
```

Comment:
"Be sure you have your TPM owner’s password and your disk recovery key or passphrase"
"Be sure you have your USB Security dongle GPG Admin PIN (HOTP) and your disk recovery passphrase, since measurements will be invalidated. HOTP sealing requires USB Security dongle's GPG Admin PIN, while setting a new boot defaut, and consequently renewing/chaning the Disk Unlock Key to be released by TPM will require you to type your Disk Recovery Key passphase to be setuped."

The TPM Owner passphrase is rarely needed if ever after initial TPM Ownership, even when flashing a new firmware. The TPM Owner passphrase is needed only to initialize NV space under Heads. It should not be required here.

What is required here after firmware upgrade is the GPG Admin PIN (USB Security dongle Admin PIN for HOTP supporting devices) to seal the secret inside of the GPG smartcard of supported HOTP USB Security dongle.

Please reply if more information is needed to clarify understanding.

Reply:
@tlaurion This is copy that you edited in the other branch based on copy that I wrote.  
I'm not sure I understand what you want here.  This copy seems to be advocating a more detailed explanation of the process but in other places we want less detail.  This seems inconsistent.  
I'm not sure what the difference is between line 21 :'drive encryption key' and line 25: 'disk recovery key'.  I tried to make the terms consistent at one point but multiple edits by both of us may have creating ambiguity.

Reply:
@jtmoree-github-com, just saw that the table is screwed.
Image better then words for this [section](https://jtmoree-github-com.github.io/heads-wiki/BootOptions/):
![2021-06-23-163543](https://user-images.githubusercontent.com/827570/123164663-3f022500-d441-11eb-91d6-ee26fc971a9d.png)
![2021-06-23-163602](https://user-images.githubusercontent.com/827570/123164673-41647f00-d441-11eb-9399-afc149e6df38.png)

Reply:
> I'm not sure I understand what you want here. This copy seems to be advocating a more detailed explanation of the process but in other places we want less detail. This seems inconsistent.

Its not needed for firmware post upgrade instructions:
"Be sure you have your TPM owner’s password and your disk recovery key or passphrase". What is actually needed is GPG User PIN

1- TPM owner passphrase is not needed here. It is needed to initialize the TPM and nvram space prior of sealing/unsealing. Reowning TPM is not needed after upgrading firmware. Only resealing TOTP/HOTP is needed. The 1.1- GPG Admin PIN will be needed after firmware upgrade if user is using HOTP version of the ROM, since secret is sealed both inside TPM and inside GPG smartcard by asking user to type GPG Admin PIN.
2- For the Disk Unlock Key, the Disk Recovery Key passphrase is prompted to unlock LUKS container and add the key generated by Heads into the TPM NV space (no TPM passphrase owner passphrase promt here). Then since the boot default config changed, the user will also need to type his GPG User PIN to sign /boot content.

This is the same part we are constantly arguing about for important details, since you want to give some but those needs to be the good ones otherwise I prefer them to not be given else users will be more confused, not less.

Reply:
@jtmoree-github-com 

>I'm not sure what the difference is between line 21 :'drive encryption key' and line 25: 'disk recovery key'. I tried to make the terms consistent at one point but multiple edits by both of us may have creating ambiguity.

https://jtmoree-github-com.github.io/heads-wiki/Updating/#upgrading-heads :
"The first time you install Heads, you’ll need a hardware flash programmer to be able to replace the existing vendor firmware. Subsequent upgrades can be performed via software, although you’ll probably want a hardware programmer since we don’t have a fail-safe recovery mechanism in the event of a bad flash or buggy firmware.

Additionally, flashing a new firmware will change the TPM PCRs. This will require generating a new TPM TOTP token and a new drive encryption key. Be sure you have your TPM owner’s password and your disk recovery key or passphrase available since, by design, the disk keys are not accessible to the recovery shell.

If you flash the same firmware and you keep settings, your TOTP will be valid, HOTP also, and Disk Unlock Key passphrase will still boot your system. In doubt, you can consequently reflash your firmware.

The Disk Recovery Key is the key used at OS installation for the encrypted root partition (passphrase placed in LUKS keyslot 0)."



https://osresearch.net/Updating/#upgrading-heads :
"Additionally, reflashing the firmware will change the TPM PCRs. This will require generating a new TPM TOTP token and a new drive encryption key. Be sure you have your TPM owner’s password and your disk encryption recovery key or passphrase available since, by design, the disk key is not accessible to the recovery shell."


Confusing parts:
drive encryption key = TPM LUKS decryption key per @osresearch used terminology (called Disk Unlock Key everywhere else, I think...)

Reply:
This paragraph mentions keys which are not listed in keys.md.  It is confusing to everyone which is why I tried to clarify the terminology.  The keys page (https://osresearch.net/Keys/) lists 3 keys related to disks but the docs include 10 different terms.  

# officially listed related to drive/disk

* TPM disk encryption key
* Disk recovery key
* LUKS disk encryption key

# terms used in docs

## TPM disk encryption key:

* keys.md =>1
* configuring-keys.md: => 1

## Disk recovery Key:

* keys.md => 4
* install-os => 4

## LUKS disk encryption key:

* keys.md => 1

## Disk encryption key without explicit LUKS/TPM:

* keys.md => 4
* upgrading.md => 1

## Disk encryption recovery key:

* upgradigng.md => 1

## Disk Key:

* faq.md => 2
* install-os.md => 1
* upgrading.md => 2

## TPM disk unlock key:

* keys.md => 1
* install-os.md => 1

## Disk unlock key passphrase:

* keys.md => 3

## Disk unlock key:

* keys.md => 4
* install-os.md => 4

## TPM sealed drive keys:

* heads-threat-model.md => 1

#  JT changes

After I worked on the docs it is mostly the same but a few changed

## TPM disk encryption key:  SAME

## Disk recovery Key:
* keys.md => same
* install-os => same
* upgrading.md => 2

## LUKS disk encryption key: SAME

## Disk encryption key without explicit LUKS/TPM:
* keys.md => SAME
* configuring-keys.md => SAME
* upgrading.md => 0

## Disk encryption recovery key:  NONE

## Disk Key:
* faq.md => same
* install-os.md => same
* upgrading.md => 1 (was 2)

## TPM disk unlock key: SAME

## Disk unlock key passphrase:
* keys.md => same
* upgrading.md => 1

## Disk unlock key:
* keys.md => SAME
* install-os.md => SAME
* recoveryshellmd => 1
* upgrading.md => 1

## TPM sealed drive keys: SAME

@tlaurion  Which of the listed keys in keys.md should be mentioned in the paragraph in upgrading.md?  If the keys that should be mentioned are not listed in keys.md I will add as part of this change.

Reply:
@jtmoree-github-com Sorry for the delay...
------
[TPM disk encryption key](https://osresearch.net/Keys/#tpm-disk-encryption-key)  = TPM sealed drive keys = TPM Disk Unlock Key  = TPM released Disk Unlock Key : Disk Unlock Key (different the Disk Recovery Key, which is setuped at OS installation, below)

This is the random key created under Heads, sealed under TPM NV space, which is unlocked with its associated passphrase when and only the measurements and the passphrase are valid as shown at https://osresearch.net/Keys/#disk-unlock-key-passphrase-prompt-output 

This TPM released Disk Encryption Key (Disk Unlock Key) is added into slot 1 of the LUKS header (after being [transformed](https://osresearch.net/Keys/#luks-disk-encryption-key) and stored in the LUKS header. So here again, The TPM DISK encryption key != actual key stored under LUKS header.

------
[Disk Recovery Key](https://osresearch.net/Keys/#disk-recovery-key) = (LUKS Disk encryption key, see below) = Disk Key = Disk encryption recovery key = Disk encryption key without explicit LUKS/TPM : Disk Recovery Key

This is the LUKS encryption key that was setuped during OS install with user provided Disk Recovery Key passphrase. This one is stored into LUKS header slot 0. 

This Disk Recovery Key is added into slot 0 of the LUKS header (after being [transformed](https://osresearch.net/Keys/#luks-disk-encryption-key) and stored in the LUKS header. So here again, The Disk Recovery Key != actual key stored under LUKS header.

Notes: The reason I insist on slots is because a misinformed user deciding to change the LUKS passphrase through a cryptsetup manually, without specifying which slot to use, will have cryptestup default behavior which is to create the new key that was into slot 0 and put it into slot 1. Heads takes care of this by specifying which [slots it interacts with.](https://github.com/osresearch/heads/blob/c3b0bd6ffbe816430dd41ef54e649af52ed1ff3b/initrd/bin/kexec-seal-key#L66-L75)

Now.

Disk Unlock Key passphrase != Disk Recovery Key passphrase.
Disk Key != Disk passphrase.

One of the problem is consistency, the other is definition, and the third is passing a message to users.
I prefer Disk Recovery Key and Disk Unlock Key and talking about associated passphrases then reusing the same words but adding complications like when using TPM released Disk encryption key. My point here is that they should all be the same. Using the same words that are used under https://osresearch.net/Keys might be the simplest to do here and making sure those terms are the same everywhere. Changing them everywhere to proper terms, after proper debate, should be done at a later time. This is really confusing for everyone, as you can see here. My version of this ROM is using Disk Unlock Key and Disk Recovery Key passphrase. We will be able to switch to those terms later on after consensus.
-------
[LUKS disk encryption key](https://osresearch.net/Keys/#luks-disk-encryption-key) = LUKS disk encryption key!
This entree explains what a LUKS key actually is.

Reply:
I have reverted the paragraph that you quoted.  It and all of the terminology issues discussed are now out of scope for this pull request.